tla-io/src/main/scala/at/forsyte/apalache/io/tnt/TntIR.scala

```diff
@@ -0,0 +1,346 @@
+  implicit val rw: RW[TntDef] = RW.merge(Def.rw, Var.rw, Const.rw, Assume.rw, TypeDef.rw)
+
+  trait WithOptionalTypeAnnotation {
+    val typeAnnotation: Option[TntType]
+  }
+
+  trait WithTypeAnnotation {
+    val typeAnnotation: TntType
+  }
+
+  /**
+   * A user-defined operator that is defined via one of the qualifiers: val, def, pred, action, or temporal. Note that
+   * TntOpDef does not have any formal parameters. If an operator definition has formal parameters, then `expr` should
+   * be a lambda expression over those parameters.
+   */
+  @key("def") case class Def(
+      id: Int,
+      /** definition name */
+      name: String,
+      /** definition qualifier: 'val', 'def', 'action', 'temporal' */
+      // Should it be a sum type?
+      qualifier: String,
+      /** expression to be associated with the definition */
+      expr: TntEx,
+      typeAnnotation: Option[TntType] = None)
+      extends TntDef with WithOptionalTypeAnnotation {}
+  object Def {
+    implicit val rw: RW[Def] = macroRW
+  }
+
+  @key("const") case class Const(
+      id: Int,
+      /** name of the constant */
+      name: String,
+      typeAnnotation: TntType)
+      extends TntDef with WithTypeAnnotation {}
+  object Const {
+    implicit val rw: RW[Const] = macroRW
+  }
+
+  @key("var") case class Var(
+      id: Int,
+      /** name of the variable */
+      name: String,
+      typeAnnotation: TntType)
+      extends TntDef with WithTypeAnnotation {}
+  object Var {
+    implicit val rw: RW[Var] = macroRW
+  }
+
+  @key("assume") case class Assume(
+      id: Int,
+      /** name of the assumption, may be '_' */
+      name: String,
+      /** an expression to associate with the name */
+      assumption: TntEx)
+      extends TntDef {}
+  object Assume {
+    implicit val rw: RW[Assume] = macroRW
+  }
+
+  /**
+   * TntTypeDefs represent both type aliases and abstract types
+   *
+   *   - Abstract types do not have an associated `type`
+   *   - Type aliases always have an associated `type`
+   */
+  @key("typedef") case class TypeDef(
+      id: Int,
+      /** name of a type alias */
+      name: String,
+      /**
+       * type to associate with the alias (none for uninterpreted type)
+       *
+       * Type aliases have `Some(type)` while abstract types have `None`
+       */
+      // TODO need special instruction for this conversion
+      typ: Option[TntType])
+      extends TntDef {}
+  object TypeDef {
+    implicit val rw: RW[TypeDef] = macroRW
+  }
+}
+
+/**
+ * TNT Types
+ *
+ * Note that TNT types have an optional associated source ID. We ignore that in our representation currently, as we have
+ * no use for source IDs on types.
+ */
+sealed trait TntType
+
+object TntType {
+
+  implicit val rw: RW[TntType] = RW.merge(
+      BoolT.rw,
+      IntT.rw,
+      StrT.rw,
+      ConstT.rw,
+      VarT.rw,
+      SetT.rw,
+      SeqT.rw,
+      FunT.rw,
+      OperT.rw,
+      TupleT.rw,
+      RecordT.rw,
+      UnionT.rw,
```

Comment:
Same here. How much collision are we going to cause with other names e.g. `BoolT1`? For the time being, I would prefix all of them with `Tnt`.

Reply:
The intended way to use this would be to avoid importing the enclosing object. But as with the value constructors, I'll use Hungarian notation.

Reply:
Fixed in aa5a4ea946a75749927d6c6fc88975e530f68f53